Fix : Connection Error

kim-howard opened the pull request:
Splash 가 추가됨에 따라서 rootViewcontroller가 splash로 변경됨에 따라
AppDelegate의 scrapViewController로의 연결이 끊어진 현상을 해결하였습니다! 
AppDelegate가 아닌 TabBarController에서 연결을 설정하였습니다!

kim-howard committed:
Fix : Connection Error

Fix
Fix connection Error with Scrap Tab
because splash controller added as a rootViewController

godpp commented on tree/tree/Controller/Scrap/ScrapTabBarController.swift:
```diff
@@ -25,10 +25,14 @@ class ScrapTabBarController: UITabBarController {
+            let viewControllers = self.viewControllers,
+            viewControllers.count >= 3,
+            let navigationController = viewControllers[2]
+                as? UINavigationController,
```
이렇게 줄바꿈 하신 이유가 있나요?? 🤔

kim-howard replied:
네이밍 변경하다가 실수한것 같습니다 😭 컨벤션에 맞춰서 다시 작성하겠습니다!

godpp reviewed (commented):
버그 수정 감자합니다! 😄

kimhyeri reviewed (approved):
🤗

yagom commented on tree/tree/Controller/Scrap/ScrapTabBarController.swift:
```diff
@@ -25,10 +25,14 @@ class ScrapTabBarController: UITabBarController {
-        scrapViewController.setupScrapBadgeValue()
+            let viewControllers = self.viewControllers,
+            viewControllers.count >= 3,
+            let navigationController = viewControllers[2]
```
과연 이 방법이 최선일까요?

kim-howard replied:
어떤 부분을 말씀하신 것인지는 모르겠지만 저는 두가지 부분에서 다시 생각해봤습니다. 

첫번째로 저번에 튜터님이 리뷰때 말씀하신 것 처럼 사실상 안에 있는 인덱스도 있는지 확인해야 한다고 하셔서 카운트를 통해서 해결 해야겠다고 생각했습니다. 하지만 일일이 count 로 세서 해결하는 것 보다 배열 안에 인덱스가 있는지 확인하는 익스텐션을 만들어야 겠다는 생각이 들어 익스텐션을 생성하였습니다.

두번째로 초기화 과정은 viewDidLoad 에서 해야한다고 생각하는데 해당 함수내의 tabarcontroller에서 scrapViewController를 확인할 수 없어서(메모리에 로드되는 타이밍이 다른 것 같은데 이 부분을 해결하지 못하였습니다 ㅠㅠ ) viewDidAppear에서 해결하였습니다. 하지만 코드의 의미 상 완벽하지 않다는 생각이 들었고 현재 DidLoad에서는 값을 초기화하고 DidApear에서는 함수를 호출하는 방식으로 코드를 작성하였습니다.

말씀하신 방식이 맞는지에 대해서 피드백 주시면 감사하겠습니다!! 👍

yagom replied:
배열의 안전하지 못한 사용에 대해 언급한 것은 아니었습니다.
제 의도는... 만약 탭바의 뷰 컨트롤러 순서가 바뀐다면 이 코드도 다시 수정되어야 하지 않을까에서 출발합니다.
만약 우리가 찾아가고자 하는 내비게이션 컨트롤러가 내비게이션 컨트롤러를 상속받은 특정 클래스라면 탭바컨트롤러의 뷰 컨트롤러 배열에서 타입을 통해 찾아낼 수 있을거라고 생각했고, 상속이 번거롭다면 빈 프로토콜을 만들어서 해당 내비게이션 뷰 컨트롤러에 채택한다면 인덱스 변화에도 유연하게 해당 내비게이션 컨트롤러를 찾아갈 수 있을 것이라고 생각했습니다.

```swift
protocol ABC { }

extension MyNavi: ABC { }

tabbarcontroller.viewControllers.filter { $0 is ABC }
```
이런 생각을 해보았습니다.
또, 내비게이션 컨트롤러의 특정 뷰 컨트롤러를 찾아가는 것도 마찬가지로 
```swift
navicontroller.viewControllers.filter { $0 is SomeViewController }
```
이렇게 찾아올 수 있지 않을까요?

kim-howard replied:
제가 생각하지 못한 방식 이였습니다! 새로운 관점에서 접근할 수 있는 좋은 기회를 잊지 않으려 변경한 후에 푸시 날렸습니다 감사합니다 🎊 😃

kim-howard committed:
Refactor : avoid temporary solution

Refactor
Add array extension for checking index
And put function in place

godpp reviewed (approved):
고생하셨습니다! 👍 👍 👍

kim-howard committed:
Refactor : think flexible

Refactor
think if tabbarcontroller is changed in some way like change orders of viewcontrollers
refactor code with flexibility

yagom reviewed (approved):
👍
